### ISSUE TITLE
Nested portals

### ISSUE THREAD

**Issue by lazlo-bonin** [LinusBorg/portal-vue#102]
How can I have a portal inside a portal, both pointing to the same portal-target?

I'm using PortalVue to display all my modal dialogs in a top-level div to avoid z-ordering issues.

For this, I have a `my-dialog` component that defines a portal at its root:

    <template>
        <portal to="dialogs">
            <div class="dialog-mask">
                <div class="dialog-container">
                    ...

And a single `dialogs` portal-target:

    <div id="app">
        ...
        <portal-target name="dialogs" multiple></portal-target>
        ....
    </div>

Many of my components "include" a dialog (e.g. for their related messages, etc.). This leads to the case where, sometimes, a dialog can include another dialog (e.g. when an action button in the top-level dialog defines its own nested success/failure dialogs as part of its template).

This leads to very weird behaviour in PortalVue. Sometimes, the component that defines the nested dialog will stop rendering when it gets opened. I created the following test case to try isolating the issue:

    <my-button @click="showOuter = true">Open Outer Dialog</my-button>
    	
    <my-dialog v-if="showOuter" title="Outer" @close="showOuter = false">
        <my-button @click="showInner = true">Open Inner Dialog</my-button>
		
        <my-dialog v-if="showInner" title="Inner" @close="showInner = false">
            Inner
        </my-dialog>
    </my-dialog>

This led to the following error (and infinite recursion) whenever the inner dialog is opened:

    [Error] [Vue warn]: You may have an infinite update loop in a component render function.
    found in
    ---> <PortalTarget>
               <Root>
           warn (app.js:24505)
           flushSchedulerQueue (app.js:26884)
           (anonymous function) (app.js:25738)
           flushCallbacks (app.js:25659)
           promiseReactionJob

How would you advise to setup PortalVue in this case?

**Comment by LinusBorg** [comment]
Hm, that's  tricky one. Neve thought of that use case.

The infinite loop is because you are essentially sending content from within the target to itself over and over.

I can't think of a workaround for you on the spot. Maybe this is fixable in the lib if we somehow can make a `<portal>` realize it's itself being rendered in the `<portal-target>`its sending its content to.

For a secon I thought a solution based on `provide/inject` would work, but it won't, since the `$parent` chain of the portal won't contain the `<portal-target >`. But maybe I can make that $parent redirection optional?

Anyhow, I have to think about this more deeply. As a heads up I will tell you that I'm in the middle of preparing for a move to a new flat, as well as some trips, so I probably won't find time to work on this in the coming weeks until March.

**Comment by lazlo-bonin** [comment]
Thanks for the quick reply; unfortunately this is an urgent issue for our project which is nearing deployment in a few days. If you can think of any workaround, please let me know!

**Comment by LinusBorg** [comment]
I will think about it a bit, but will be on by brother's bachelor trip this weekend, so there won't be much time. 

As I indicated previously, a not so clean but maybe doable workaround would be to provide a sequence of targets (like z-indexes) and tell the dialog compontent to which one to portal with a prop.

**Comment by LinusBorg** [comment]
Maybe you could even automate that by having your portal check how many other dialogs are in its $parent chain and determine the right target accordingly

**Comment by LinusBorg** [comment]
Example: Given these targets:
```html
<portal-target name="level-0" />
<portal-target name="level-1" />
<portal-target name="level-2" />
```
You could dynamically determine how many dialogs you are nested by doing this in your dialog:
```javascript
function getLevel() {
  let levels = 0
  let parent = this.$parent
  while (parent) {
  if (parent.$options.name === 'dialog') {
    level++
  }
  parent = parent.$parent
  }
return level
}
```
```html
<portal :to="`level-${getLevel()}`">
```

Obviously if you have different kinds of controls that can be nested with portals inside them (dropdowns ...) the detection via `parent.$options.name` would have take this into account, but I hope this gets you on a workable path until we find a proper solution.

**Comment by lazlo-bonin** [comment]
Clever workaround! Tried it, but it seems like parent becomes the portal target's parent itself once the dialog is moved by PortalVue. Therefore, `getLevel()` always returns 0 and the problem persists.

**Comment by LinusBorg** [comment]
Hm I didn't expect that, won't be able to test this weekend, unfortunately. 

I thought it should work because I explicitly took care in the li  that $parent inside a portal points to the portal's parent, not the target's.

**Comment by lazlo-bonin** [comment]
Other workaround I tried: having a global `dialogLevel` variable that gets incremented when the dialog component is mounted, and decremented when it gets destroyed. However, it seems that the component gets mounted/destroyed when being moved by PortalVue, and thus going back to the infinite loop...

**Comment by LinusBorg** [comment]
Oh, now I see what's happening... of course. The template with all nested slots is fully turned into vnodes within the parent component's context, and then the vnodes are passed down through the slots. So this won't work indeed.

Still should work (or could, don't want to sound too secure) if:

* you set the different targets manually on the dialogs' portals, or
* wrap the portal in another small wrapper component and do the above calculation in there:

```html
<wrapped-portal to="level">

</wrapped-portal>
```
```javascript
// wrapped-portal.js
export default {
  render(h) {
    props: ['to'],
    return h('portal', { 
      props: {
        ...this.$attrs, 
        to: this.to  + '-'  + this.getLevel 
      } 
    }
  },
  methods: {
    getLevel() { ... }
  }
}
```

Might work better...

Anyway, I won't be able to do much more until sunday, maybe even tuesday, sorry :/

**Comment by lazlo-bonin** [comment]
See my edit, I found a working, automated workaround :) Thanks for all the help!

**Comment by LinusBorg** [comment]
Oh, glad it works! Good luck with the deadline!

**Comment by andrey-hohlov** [comment]
I created Modal with a plugin and have this.$modal in all components.
$modal is reactive.
Each Modal component adding to the list when shows and removing from the list when hide.

Portal created dynamically:
```
<portal-target
  v-for="level in $modal.count + 1"
  :key="level"
  :name="`modals-level-${level}`"
  multiple
/>
```
$modal.count is reactive and = opened modals list length.

And anywhere in the application, I use:
```
<portal :to="`modals-level-${$modal.getNextLevel()}`">
  <modal/>
</portal>
```
$modal.getNextLevel() returns counter that increasing when modal showing and descreasing when hiding (not reactive).

That works. Modal mounted in modal knows next level portal, new portal always ready for the new level.

But I found that ugly anyway :(

**Comment by andrey-hohlov** [comment]
I make this component.
It's cut-version, in full I mark the previous modal as "freezed" to remove the background and wrap all in animations with anime.js.

<details>
  <summary>Example</summary>
  <img src="https://user-images.githubusercontent.com/7358151/35763482-cdb29e5a-08df-11e8-8a36-164b8def3e95.png"/>
</details>
<p><p>

There is some strange moments like `this.mountedComp.$el.parentNode.removeChild(this.mountedComp.$el);`
But only with it it working (include hot module replacement).

Modal.vue
```
<script>
  import Vue from 'vue';
  import { PortalTarget } from 'portal-vue';

  const showing = [];

  export default {
    props: {
      show: {
        type: Boolean,
        default: false,
      },
    },
    data() {
      return {
        isShow: false,
        zIndex: 1000,
        pid: 'modal',
      };
    },
    watch: {
      show(val) {
        this.toggle(!!val);
      },
      $route() {
        this.toggle(false);
      },
    },
    beforeMount() {
      this.pid = `modal-${Date.now()}`;
      const el = document.createElement('div');
      document.body.appendChild(el);
      const newTarget = new Vue({
        ...PortalTarget,
        parent: this,
        propsData: {
          name: this.pid,
          tag: el.tagName,
          slim: true,
        },
      });
      newTarget.$mount(el);
      this.mountedComp = newTarget;
    },
    beforeDestroy() {
      if (this.mountedComp) {
        this.mountedComp.$el.parentNode.removeChild(this.mountedComp.$el);
        this.mountedComp.$destroy();
        this.mountedComp = null;
      }
      this.toggle(false);
    },
    methods: {
      toggle(state) {
        if (this.locked) return;

        if (state && !this.isShow) {
          showing.push(this);
          this.isShow = true;
          this.zIndex = 1000 + showing.length;
          this.$emit('show');
          return;
        }

        if (!state && this.isShow) {
          const index = showing.indexOf(this);
          if (index < 0) return;
          showing.splice(index, 1);
          this.isShow = false;
          this.$emit('hide');
        }
      },
    },
  };
</script>

<template>
  <portal :to="pid">
    <div
      class="modal"
      v-if="isShow"
      :style="{ zIndex }"
    >
      <div class="modal__overlay"/>
      <div class="modal__wrap">
        <div class="modal__content">
          <button
            type="button"
            class="modal__close"
            @click.prevent="toggle(false)"
          >
            Close
          </button>
          <div class="modal__body">
            <slot/>
          </div>
        </div>
      </div>
    </div>
  </portal>
</template>

<style lang="scss" type="text/scss">
  .modal {
    position: fixed;
    z-index: 1000;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    text-align: center;
    overflow-y: auto;
    white-space: nowrap;
    font-size: 0;

    &::before {
      content: '';
      display: inline-block;
      height: 100%;
      width: 0;
      vertical-align: middle;
    }

    &__overlay {
      position: fixed;
      z-index: 950;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      background: rgba(#000, .8);
    }

    &__wrap {
      position: relative;
      z-index: 950;
      display: inline-block;
      vertical-align: middle;
      margin: 0 auto;
      text-align: initial;
      white-space: normal;
      max-width: 100%;
      font-size: 14px;
      line-height: 1.4em;
    }

    &__content {
      position: relative;
      margin: 0 auto;
      width: 600px;
      max-width: 100%;
      background: #fff;
    }

    &__body {
      position: relative;
      padding: 20px;
    }
  }
</style>

```

**Comment by LinusBorg** [comment]
I'm not surei
 I understand what has to do with IOS issue of nested portals, can you explain?

**Comment by andrey-hohlov** [comment]
@LinusBorg you ask me?

**Comment by LinusBorg** [comment]
Yes

**Comment by andrey-hohlov** [comment]
Sorry if my message not clear. I just show my component, that solve problem with nesting portals.

**Comment by LinusBorg** [comment]
I decided to close this feature request as it hasn't come up from anyone else so far and would be very complex to solve.

**Comment by hrakotom** [comment]
Actually, I would have a use case for this one :
- I need to apply transforms to the portal children dissociated from the parent's transform.
( On one side you would have the skeleton 'dom tree' having child portals all rendering to a unique destination portal )

It would be possible to handle this via vuex but I would lose all routing/hierarchy info. I had a glimmer of hope when I saw this lib, but this just crushed it.

Could you point me to a potential solution, so I can see if I can help ? Would adding a unique id to source portals potentially solve the problem ?

**Comment by hrakotom** [comment]
OK, named portals and a specialized container worked...
Thanks for the lib :)

**Comment by renatodeleao** [comment]
I was about to create a new issue for 3.0 but realised this existent one which is on the same subject so decide to post here. (sry for reviving the thread 😅  )

@LinusBorg What's weirder to me is that nested portals (on dialogs) always worked for me in `vue2` with `portal-vue@2` — I've used it heavily as part of my [`a11y-vue-dialog`](https://github.com/renatodeleao/a11y-vue-dialog) plugin

I've recently migrated a big codebase to `vue3` that made use of this feature and noticed that the same usages of the portal were now broken: dropdown menus with nested levels and nested dialogs stopped working (infinite loop). 

Can this be considered a regression in vue3?

- [Here's the sandbox code with `vue2/portal-vue@2` working fine](https://codesandbox.io/s/portal-vue-2-x-recursive-error-not-happening-wdrjyr?file=/src/components/FakeDialog.vue:0-1144)
- [Here's the same sandbox migrated to `vue3/portal-vue@3.0.0` infinite looping](https://codesandbox.io/s/portal-vue-3-0-recursive-error-tns3e1?file=/src/App.vue)

```
[Vue warn]: Maximum recursive updates exceeded in component <portalTarget>.
```

As a workaround, from a quick test, the native vue3 `Teleport` component seems to handle nesting without issues, but `Teleport` still isn't a full drop-in replacement for `portal-vue` yet (https://github.com/vuejs/core/issues/2015)
